Riot/Managers/URLPreviews/URLPreviewStore.swift

```diff
@@ -0,0 +1,158 @@
+// 
+// Copyright 2021 New Vector Ltd
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+import CoreData
+
+/// A cache for URL previews backed by Core Data.
+class URLPreviewStore {
+    
+    // MARK: - Properties
+    
+    /// The Core Data container for persisting the cache to disk.
+    private let container: NSPersistentContainer
+    
+    /// The Core Data context used to store and load data on.
+    private var context: NSManagedObjectContext {
+        container.viewContext
+    }
+    
+    /// A time interval that represents how long an item in the cache is valid for.
+    private let dataValidityTime: TimeInterval = 60 * 60 * 24
+    
+    /// The oldest `creationDate` allowed for valid data.
+    private var expiryDate: Date {
+        Date().addingTimeInterval(-dataValidityTime)
+    }
+    
+    // MARK: - Lifecycle
+    
+    /// Create a URLPreview Cache optionally storing the data in memory.
+    /// - Parameter inMemory: Whether to store the data in memory.
+    init(inMemory: Bool = false) {
+        // Register the transformer for the `image` field.
+        ValueTransformer.setValueTransformer(URLPreviewImageTransformer(), forName: .urlPreviewImageTransformer)
+        
+        // Create the container, updating it's path if storing the data in memory.
+        container = NSPersistentContainer(name: "URLPreviewStore")
+        
+        if inMemory {
+            container.persistentStoreDescriptions.first?.url = URL(fileURLWithPath: "/dev/null")
+        }
+        
+        // Load the persistent stores into the container
+        container.loadPersistentStores { storeDescription, error in
+            if let error = error {
+                MXLog.error("[URLPreviewStore] Core Data container error: \(error.localizedDescription)")
+            }
+        }
+    }
+    
+    // MARK: - Public
+    
+    /// Store a preview in the cache. If a preview already exists with the same URL it will be updated from the new preview.
+    /// - Parameter preview: The preview to add to the cache.
+    /// - Parameter date: Optional: The date the preview was generated.
+    func store(_ preview: URLPreviewData, generatedOn generationDate: Date? = nil) {
```

Comment:
After a refactor the call site of this becomes `store.store(...)`. Which isn't very nice, so will update!